modules/sr/robot3/arduino_devices.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import annotations
+
+import abc
+import logging
+from typing import Optional
+
+from controller import (
+    LED,
+    Robot,
+    TouchSensor,
+    DistanceSensor as WebotsDistanceSensor,
+)
+from sr.robot3.utils import map_to_range, get_robot_device
+from controller.device import Device
+from sr.robot3.randomizer import add_jitter
+from sr.robot3.output_frequency_limiter import OutputFrequencyLimiter
+
+LOGGER = logging.getLogger(__name__)
+
+
+class PinDevice(abc.ABC):
+    """
+    A device connected to a pin on the Arduino.
+    """
+    _ANALOG_MIN = 0.0  # Volts
+    _ANALOG_MAX = 5.0
+    _DEVICE_TYPE: Optional[type[Device]] = None
+    _webot_device: Optional[Device]
```

Comment:
style suggestion -- it would be nicer to separate out the webots nature of the devices from their device simulation nature. This would more easily allow for non-webots devices, such as the null device (`EmptyPin`).

One approach might be a base type which just implements the base API and has a name, another might be to use a protocol and decouple the inheritance chains of the non-webots devices.

Reply:
What we've got here works for the moment, we can always re-evaluate if we find a need for additional non-webots devices.

Reply:
What we have in #411 _works_, if that's the bar we're aiming for?
I had thought the concerns with that approach were around not sufficiently well separating out the various modelling concerns. Perhaps I'd misunderstood?

Reply:
Yes both of the PRs work, however I think this implimentation is clearer than #411 
I think Will was refering to the fact at the moment we do not have any non-webots devices apart from the EmptyPin. If this were to change we can then revisit this discussion to what best suits the new device we want to add